### PR TITLE
[stm32] PM restore console after sleep mode

### LINKED ISSUE #31284: [stm32] PM restore console after sleep mode
When testing the PowerManagement on stm32L4 or stm32WB55 through the tests/subsys/power/power_mgmt
the console is no more functional and prints spurious characters.

**Describe the solution you'd like**
To define a function in the uart_stm32.c driver (replace _device_pm_control_nop_)  
to flush the  console before going to sleep 
To restore the correct clock for the console after sleep mode

**Console: **
```

Running test suite power_management_test                                        
===================================================================             
START - test_power_idle                                                         
give way to idle thread                                                         
 PASS - test_power_idle                                                         
===================================================================             
START - test_power_state_trans                                                  
�������������������������������������������������������������������������������

```

### PR DESCRIPTION
**Pull request by FRASTM**
With this change, the clock configuration is restored when leaving the low power mode. 
Then the console of the nucleo_wb55 can operate after sleeping.

The UART executes the `uart_stm32_set_power_state` to have a clean configuration before entering the low power mode.
On wakeup, the clocks are re-configured thanks to the `stm32_clock_control_init` function called directly in the 
`pm_power_state_exit_post_ops` of each soc.
This method is more efficient than introducing a _clock_stm32_pm_control_ function in the clock driver. 

Fix #31284
Requires #32042

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

**Comment by FRASTM** [comment]
*** Booting Zephyr OS build v2.5.0-rc1-116-g9d9c34e27c13  ***                                                                           
Running test suite power_management_test                                                                                                
===================================================================                                                                     
START - test_power_idle                                                                                                                 
give way to idle thread                                                                                                                 
 PASS - test_power_idle                                                                                                                 
===================================================================                                                                     
START - test_power_state_trans                                                                                                          
 PASS - test_power_state_trans                                                                                                          
===================================================================                                                                     
START - test_power_state_notification                                                                                                   
 PASS - test_power_state_notification                                                                                                   
===================================================================                                                                     
Test suite power_management_test succeeded                                                                                              
===================================================================                                                                     
PROJECT EXECUTION SUCCESSFUL

File: drivers/serial/uart_stm32.c
Comment (tagunil) [inline drivers/serial/uart_stm32.c:59]: What's the point of this variable? Is there some weird reading optimization problem?

Comment (gmarull) [reply drivers/serial/uart_stm32.c:59]: Its scope could likely be at `while (LL_USART_IsActiveFlag_RXNE(UartInstance)) {` loop

File: drivers/clock_control/clock_stm32_ll_common.h
Comment (gmarull) [inline drivers/clock_control/clock_stm32_ll_common.h:42]: If made public it's worth adding some docstrings

File: drivers/serial/uart_stm32.c
Comment (erwango) [inline drivers/serial/uart_stm32.c:39]: Leftover ?

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c:59]: var is used to flush the rx fifo, now moved in the function itself

File: drivers/clock_control/clock_stm32_ll_common.h
Comment (FRASTM) [reply drivers/clock_control/clock_stm32_ll_common.h:42]: ok, adding some lines

**Comment by FRASTM** [comment]
Remove the commit  to` soc: arm: stm32wb low power mode for CPU2 too` . The CPU2 lowpower mode is not set here (LL_C2_PWR_SetPowerMode(LL_PWR_MODE_STOPx);
and changes after review

File: drivers/serial/uart_stm32.c
Comment (gmarull) [inline drivers/serial/uart_stm32.c:1465]: It may be part of an `} else if () {`, then just return for the rest. If any `ctrl_command` gets added, this code could still be valid but would fail due to this assertion.

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: not really, removed

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: yes, now changed

Comment (FRASTM) [reply drivers/serial/uart_stm32.c]: uart_stm32_set_power_state simplified
adding the power.c for stm32l4 : restoring the clock with `stm32_clock_control_init`

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c:1440]: simplification in  testing only new_state != DEVICE_PM_ACTIVE_STATE

File: drivers/serial/uart_stm32.c
Comment (dcpleung) [inline drivers/serial/uart_stm32.c:1394]: This does not _exactly_ set the power state, which is confusing to future readers. I suggest moving the `data->pm_state = new_state` from below to here, or rename this function to something like `uart_stm32_prep_power_transition()`.

**Comment by abel-NoelleC** [comment]
Thank you for this PR @FRASTM ! It would help me a lot if it was merged since it's required for STM32L1x power management too (when using the stop mode).

**Comment by erwango** [comment]
> Thank you for this PR @FRASTM ! It would help me a lot if this was merged since it's required for STM32L1x power management too (when using the stop mode).

@abel-NoelleC The best way to help in this direction is review, and approve when you're satisfied (in the scope on the proposal)

File: drivers/serial/uart_stm32.c
Comment (abel-NoelleC) [inline drivers/serial/uart_stm32.c:1393]: `uart_stm32_pm_control()` is now also part of this #ifdef. Could this cause problems when `CONFIG_PM_DEVICE` isn't selected?

**Review by abelNoelleC — APPROVED** [review]
If @dscpleung 's comment is resolved, and if having `uart_stm32_pm_control()` inside the `CONFIG_PM_DEVICE` bounds doesn't cause problems, this would meet our needs. 

Additionally I'd like to discuss if peripheral (in this case UART) specific GPIO configuration for power management would also fit in these drivers. Not just GPIO port configuration, but also peripheral deinitialization (for example for UART use of `LL_USART_DeInit(USARTx)` function, where peripheral clock is disabled). 
IMO not something that blocks this PR, but something to keep in mind and discuss now that the power management is being added to these peripheral device drivers.

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c:1393]: In case the PM_DEVICE is not defined, the uart_stm32_pm_control does not exist but  DEVICE_DT_INST_DEFINE get a null value for this entry (no error).

File: drivers/serial/uart_stm32.c
Comment (FRASTM) [reply drivers/serial/uart_stm32.c:1394]: However it seems to be more or less generic that the `uart_xx_pm_control` is calling a `uart_xx_set_power_state` function.
The test in the uart_stm32_set_power_state function is just checking that the UART is requested to enter a _non-DEVICE_PM_ACTIVE_STATE_

**Review by erwango — CHANGES_REQUESTED** [review]
Let's add nucleo_wb55rg to `tests/subsys/power/power_mgmt_soc/testcase.yaml`
and we're good to go.

**Comment by FRASTM** [comment]
adding a commit to include nucleo_wb55rg and nucleo_l476rg boards for the `tests/subsys/power/power_mgmt_soc`

File: soc/arm/st_stm32/stm32l4/power.c
Comment (tagunil) [inline soc/arm/st_stm32/stm32l4/power.c:12]: Why are these two includes needed here, but not in stm32wb/power.c? What's the difference?

Comment (FRASTM) [reply soc/arm/st_stm32/stm32l4/power.c:12]: The reason is that the clock_stm32_ll_common_.h   includes the LL_RCC_MSI_Disable and LL_UTILS_PLLInitTypeDef
for the stm32l4 soc serie (which is not the case with the wb55)

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c:12]: Sorry, but I don't see any includes at all in clock_stm32_ll_common.h.

**Comment by FRASTM** [comment]
Add more target for the tests/subsys/power/power_mgmt
Fix compilation build error with tests/kernel/profiling/profiling_api by adding a weak `pm_policy_low_power_devices`, for each stm32 soc (in soc/power.c) in case the policy is not available.

File: soc/arm/st_stm32/stm32l4/power.c
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l4/power.c:12]: That's why they are in the power.c.

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c:12]: Yes, but then how exactly does stm32wb/power.c work without them? Is it supposed to be this way? Though it's definitely not a blocker.

**Review by erwango — CHANGES_REQUESTED** [review]

File: soc/arm/st_stm32/stm32l4/power.c
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l4/power.c:11]: correct

**Comment by FRASTM** [comment]
adding the `include <stm32wbxx_ll_utils.h> for power.c of the stm32l4 and stm32wb, in case the PLL is enable here.

File: soc/arm/st_stm32/stm32l4/power.c
Comment (erwango) [reply soc/arm/st_stm32/stm32l4/power.c:11]: It should work the the same as it works on L4 with PLL. But point is main BLE config doesn't use PLL, meaning PLL is not a likely config on WB.

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c:11]: @erwango thanks for the clarification!
@FRASTM wouldn't it be more consistent to include <stm32l4xx_ll_utils.h> instead of the generic one?

File: drivers/serial/uart_stm32.c
Comment (erwango) [inline drivers/serial/uart_stm32.c:1438]: Nit: Indentation not correct here.

File: tests/subsys/power/power_mgmt_soc/testcase.yaml
Comment (erwango) [inline tests/subsys/power/power_mgmt_soc/testcase.yaml:4]: Only nucleo_wb55rg nucleo_l476rg supported right now.

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

File: drivers/serial/uart_stm32.c
Comment (dcpleung) [reply drivers/serial/uart_stm32.c:1394]: Only `uart_nrfx_uart.c` is setting the field in device struct outside of `uart_*_set_power_state()`. The other two drivers which have PM enabled, `uart_cc13xx_cc26xx.c` and `uart_nrfx_uarte.c`, are doing this inside the setter function. I still think it is better to actually set the state inside the setter function to be less confusing.

**Comment by FRASTM** [comment]
rebase

File: soc/arm/st_stm32/stm32l4/power.c
Comment (FRASTM) [reply soc/arm/st_stm32/stm32l4/power.c]: yes, changing to `stm32l4xx_ll_utils.h`

Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l4/power.c]: Can someone explain why the non-generic libraries would be more consistent? (Since I also have opened #32322)

Comment (tagunil) [reply soc/arm/st_stm32/stm32l4/power.c]: > Can someone explain why the non-generic libraries would be more consistent? (Since I also have opened #32322)

My point was that all the other similar includes in soc.c files across the series are non-generic, so it would be better to do the same, or to change all of them to generic ones at once.

Comment (abelNoelleC) [reply soc/arm/st_stm32/stm32l4/power.c]: Aaaah alright, thanks!

**Comment by FRASTM** [comment]
rebase

File: drivers/serial/uart_stm32.c
Comment (abelNoelleC) [reply drivers/serial/uart_stm32.c:1391]: I'm personally still pro at least renaming the function if you want it to keep it separate.

**Comment by erwango** [comment]
@dcpleung can you have a new look ?

**Review by dcpleung — APPROVED** [review]

**Pull request merged**
